scripts.js

```diff
@@ -1,348 +1,384 @@
-  };
+	async function loadSolution(user, year, day, part) {
+		const url = user.solutionUrl(year, day, part);
+		const preEl = document.createElement("pre");
+		const codeEl = document.createElement("code");
+		const lang = user.lang(year, day);
+		const encoding = ("encoding" in user ? user.encoding(day) : null) || "utf-8";
+		const decoder = new TextDecoder(encoding);
+		try {
+			const result = url ? await fetch(url) : null;
+			if (!result.ok)
+				throw new Error("No solution yet");
+			const buffer = await result.arrayBuffer();
+			const code = decoder.decode(buffer);
+			await Prism.plugins.autoloader.loadLanguages(
+				lang,
+				() => {
+					codeEl.innerHTML = Prism.highlight(code, Prism.languages[lang], user.lang);
+				}
+			);
+		} catch (err) {
+			codeEl.textContent = err.message;
+		}
+		preEl.appendChild(codeEl);
+		document.getElementById("preview").innerHTML = "";
+		document.getElementById("preview").appendChild(preEl);
+	}
 
-  async function loadSolution(user, day, part) {
-    const url = user.solutionUrl(day, part);
-    const preEl = document.createElement("pre");
-    const codeEl = document.createElement("code");
-    const lang = user.lang(day);
-    const encoding = ("encoding" in user ? user.encoding(day) : null) || "utf-8";
-    const decoder = new TextDecoder(encoding);
-    try {
-      const result = url ? await fetch(url) : null;
-      if (!(result?.ok ?? false)) {
-        throw new Error("No solution yet");
-      }
-      const buffer = await result.arrayBuffer();
-      const code = decoder.decode(buffer);
-      await Prism.plugins.autoloader.loadLanguages(
-        lang,
-        () => {
-          codeEl.innerHTML = Prism.highlight(code, Prism.languages[lang], user.lang);
-        }
-      );
-    } catch (err) {
-      codeEl.textContent = err.message;
-    }
-    preEl.appendChild(codeEl);
-    document.getElementById("preview").innerHTML = "";
-    document.getElementById("preview").appendChild(preEl);
-  }
+	function maxDay(selYear) {
+		return selYear < currentYear ? 25 : Math.min(new Date(Date.now()).getDate(), 25);
+	}
 
-  function render(strings, ...values) {
-    const html = String.raw({ raw: strings }, ...values);
-    const el = document.createElement('div');
-    el.innerHTML = html;
-    return el.children[0];
-  }
+	function render(strings, ...values) {
+		const html = String.raw({raw: strings}, ...values);
+		const el = document.createElement('div');
+		el.innerHTML = html;
+		return el.children[0];
+	}
 
-  function selectUserByIndex(index) {
-      state.index = index;
-      const el = document.querySelector(`#users .list-group-item:nth-of-type(${index + 1})`);
-      document.querySelector(".list-group-item.active").classList.remove("active");
-      el.classList.add("active");
-      loadSolution(users[state.index], state.day, state.part);
-  }
+	function selectUserByIndex(index) {
+		document.querySelector(`#users .list-group-item:nth-of-type(${index + 1})`).click();
```

Comment:
Dieses Refactoring sorgt dafür, dass du über die Key-Handler DOM-Interaktionen simulierst, die wiederum von Klick-Listenern behandelt werden müssen. 

Das führt zu extra Indirektionen, erhöht die Komplexität und hat im schlimmsten Fall einen negativen Einfluss auf die Performance (durch die Event-Kaskade). 

Die Idee dieser Funktionen war, dass es mehrere Events gibt (z.B. Klick auf die Liste & bestimmte Tastenkombinationen), die exakt das gleiche machen sollen. Ergo wurde die Funktionalität in diese Funktionen ausgelagert, die wiederum in den jeweiligen Handlern für die genannten Events aufgerufen werden.

Reply:
> Dieses Refactoring sorgt dafür, dass du über die Key-Handler DOM-Interaktionen simulierst, die wiederum von Klick-Listenern behandelt werden müssen.

Das war bewusst, da es die kürzeste Art war, das ein und ausblenden der Tage bei vergangenen Jahren zu erreichen.

> Das führt zu extra Indirektionen, erhöht die Komplexität und hat im schlimmsten Fall einen negativen Einfluss auf die Performance (durch die Event-Kaskade).

Ein click-event wird jedoch bei einer website dieser größe keinen merkbaren Effekt auf Geschwindigkeit haben, und selbst wenn doch, die Millisekunde hat man Zeit.

> Die Idee dieser Funktionen war, dass es mehrere Events gibt (z.B. Klick auf die Liste & bestimmte Tastenkombinationen), die exakt das gleiche machen sollen. Ergo wurde die Funktionalität in diese Funktionen ausgelagert, die wiederum in den jeweiligen Handlern für die genannten Events aufgerufen werden.

Das Konzept dupliziert aber vierfach Code. Kein schöner Stil.